{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "273083b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing libraries\n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt \n",
    "import pandas as pd \n",
    "import GEOparse\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_rows', 100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef2c0dd2",
   "metadata": {},
   "source": [
    "### To-do List\n",
    "\n",
    "* Entrega Parcial\n",
    "    * Construir df final (ok)\n",
    "    * Diferenciar e agrupar pacientes leves e medios/graves pelo sledai (ok)\n",
    "    * Analise de expressao diferencial\n",
    "    * Montar a rede de expressão diferencial e analisar\n",
    "\n",
    "* Neutrofilos -> Velocidade da resposta imune (reduz vies da analise do médico)\n",
    "    * -> Encontrar genes mais responsáveis pelo aumento no perecentual de neutrofilos\n",
    "    * -> Selecionar os genes mais relevantes para o aumento dos neutrofilos\n",
    "    * -> Correlação entre nivel de expressao genica e resposta imune"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4a73ac9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import GEOparse\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "class GEOParser:\n",
    "    \"\"\"\n",
    "    Parser for GEO datasets using GEOparse.\n",
    "    First version specific for GSE121239 (Lúpus).\n",
    "\n",
    "    Parameters\n",
    "    -----------\n",
    "    geo_id : str\n",
    "        GEO Series ID (e.g., 'GSE121239').\n",
    "    destdir : str\n",
    "        Directory to download GEO data.\n",
    "    gene_id_cols : list of str, optional\n",
    "        Columns to use for gene ID mapping. Default is ['ID', 'Gene Symbol', 'Gene Title'].\n",
    "    meta_map : dict, optional\n",
    "        Mapping of metadata fields to their index in characteristics_ch1. \n",
    "    \"\"\"\n",
    "    def __init__(self, geo_id, destdir, gene_id_cols=None, meta_map=None):\n",
    "        self.geo_id = geo_id\n",
    "        self.destdir = destdir\n",
    "        self.gene_id_cols = gene_id_cols or ['ID', 'Gene Symbol', 'Gene Title']\n",
    "        self.meta_map = meta_map\n",
    "        self.gse = None\n",
    "        self.gpl = None\n",
    "\n",
    "    def load(self):\n",
    "        \"\"\"\n",
    "        Load GEO Series data.\n",
    "        \"\"\"\n",
    "        self.gse = GEOparse.get_GEO(geo=self.geo_id, destdir=self.destdir, silent=True)\n",
    "\n",
    "    def try_cast(self, val, typ):\n",
    "        \"\"\"\n",
    "        Try to cast a value to a specified type, return NaN on failure.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        val : any\n",
    "            Value to cast.\n",
    "        typ : type\n",
    "            Target type for casting.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            return typ(val)\n",
    "        except (ValueError, TypeError):\n",
    "            return np.nan\n",
    "\n",
    "    def select_main_gene(self, gene_str, separator='///'):\n",
    "        \"\"\"\n",
    "        Select the main gene from a string of gene symbols separated by a given separator.\n",
    "        Parameters\n",
    "        ----------\n",
    "        gene_str : str\n",
    "            String containing gene symbols separated by the specified separator.\n",
    "        separator : str, optional\n",
    "            Separator used in the gene string. Default is '///'.\n",
    "        Returns\n",
    "        -------\n",
    "        main_gene : str or np.nan\n",
    "            The selected main gene symbol, or NaN if input is NaN.\n",
    "        \"\"\"\n",
    "        if pd.isna(gene_str):\n",
    "            return np.nan\n",
    "        # Lista de genes encontrados\n",
    "        genes = [g.strip() for g in gene_str.split(separator)]\n",
    "\n",
    "        # Prioriza genes que não começam com LOC (Locus genômico)\n",
    "        known_genes = [g for g in genes if not g.startswith('LOC')]\n",
    "        if known_genes:\n",
    "            return known_genes[0]\n",
    "        return genes[0]  # Se só tem LOC, retorna o primeiro\n",
    "\n",
    "    def parse(self):\n",
    "        \"\"\"\n",
    "        Parse the loaded GEO Series data.\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "        gene_data: pd.DataFrame\n",
    "            Dataframe with Gene expression data.\n",
    "        patient_data: pd.DataFrame\n",
    "            Dataframe with Patient metadata.\n",
    "        \"\"\"\n",
    "        if self.gse is None:\n",
    "            self.load()\n",
    "\n",
    "        # 1. Cria listas com metadados e expressao dos paciente e expressao genica \n",
    "        patient_meta_list = []\n",
    "        gene_data_list = []\n",
    "\n",
    "        # 2. Itera sobre os GSMs para extrair metadados e dados de expressão\n",
    "        for gsm_name, gsm in self.gse.gsms.items():\n",
    "\n",
    "            # 2.1. Extrai metadados do paciente\n",
    "            meta = gsm.metadata['characteristics_ch1']\n",
    "            patient_meta = {'sample_code': gsm_name}\n",
    "\n",
    "            # 2.2. Parse de metadados de acordo com o mapeamento fornecido\n",
    "            for key, idx in self.meta_map.items():\n",
    "                # 2.3 . Extrai valor da lista de metadados e faz cast apropriado\n",
    "                val = meta[idx].split(':')[1].strip() if len(meta) > idx else None\n",
    "                if key == 'sledai':\n",
    "                    val = self.try_cast(val, int)\n",
    "                elif key == 'visit_date':\n",
    "                    val = self.try_cast(val, pd.Timestamp)\n",
    "                elif key == 'neutrophil_percent':\n",
    "                    val = self.try_cast(val, float)\n",
    "                # 2.4. Adiciona informação ao dicionario de metadados do paciente\n",
    "                patient_meta[key] = val\n",
    "            \n",
    "            patient_meta_list.append(patient_meta)\n",
    "\n",
    "            # 2.5. Extrai dados de expressão gênica\n",
    "            gene_df = gsm.table[['ID_REF', 'VALUE']].copy()\n",
    "            gene_df['sample_code'] = gsm_name\n",
    "            gene_data_list.append(gene_df)\n",
    "\n",
    "        # 3. Concatena dados de expressão gênica em um único DataFrame\n",
    "        gene_data = pd.concat(gene_data_list, ignore_index=True)\n",
    "\n",
    "        # 4. GPLS: plataforma com anotação de genes, incluindo nome e símbolo\n",
    "        self.gpl = self.gse.gpls[list(self.gse.gpls.keys())[0]].table\n",
    "\n",
    "        # 5. Recupera nome e título do gene com base no ID\n",
    "        gene_data = gene_data.merge(self.gpl[self.gene_id_cols], left_on='ID_REF', right_on='ID', how='left')\n",
    "        gene_data = gene_data.rename(columns={'VALUE': 'gene_expression_value', 'Gene Symbol': 'gene_symbol', 'Gene Title': 'gene_title'})\n",
    "        gene_data = gene_data[['sample_code', 'ID', 'gene_symbol', 'gene_title', 'gene_expression_value']]\n",
    "\n",
    "        # 6. Metadados do paciente em dataframe\n",
    "        patient_data = pd.DataFrame(patient_meta_list)\n",
    "        patient_data['patient_id'] = patient_data['patient_id'].replace({'NA': np.nan})\n",
    "        patient_data = patient_data.sort_values(by=['patient_id', 'visit_date'])\n",
    "\n",
    "        return gene_data, patient_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8cb4d5e",
   "metadata": {},
   "source": [
    "## Loading Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "31b109ec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cs331285/anaconda3/envs/ml/lib/python3.13/site-packages/GEOparse/GEOparse.py:401: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  return read_csv(StringIO(data), index_col=None, sep=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "meta_map = {\n",
    "    'state': 0,\n",
    "    'patient_id': 1,\n",
    "    'sledai': 2,\n",
    "    'visit_date': 3,\n",
    "    'neutrophil_percent': 5\n",
    "}\n",
    "\n",
    "# Cria objeto parser\n",
    "parser = GEOParser(geo_id=\"GSE121239\", destdir=\"../../data/raw/\", meta_map=meta_map)\n",
    "\n",
    "# Parse dos dados e armazena em tabelas\n",
    "gene_data, patient_data = parser.parse()\n",
    "\n",
    "# Mantem o \"gene principal\" em casos de genes separados por '///'\n",
    "gene_data['main_gene'] = gene_data['gene_symbol'].apply(parser.select_main_gene)\n",
    "\n",
    "# Remove valores de microarray (Affymetrix) -> Controle de qualidade de experimento\n",
    "gene_data = gene_data[~gene_data['ID'].str.startswith('AFFX')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a65024bf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_code</th>\n",
       "      <th>state</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sledai</th>\n",
       "      <th>visit_date</th>\n",
       "      <th>neutrophil_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>GSM3428415</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>4</td>\n",
       "      <td>2009-10-08</td>\n",
       "      <td>65.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>GSM3428416</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>2</td>\n",
       "      <td>2010-01-11</td>\n",
       "      <td>61.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>GSM3428417</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>6</td>\n",
       "      <td>2010-03-29</td>\n",
       "      <td>57.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>GSM3428418</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1041</td>\n",
       "      <td>2</td>\n",
       "      <td>2009-09-24</td>\n",
       "      <td>73.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>GSM3428419</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1041</td>\n",
       "      <td>10</td>\n",
       "      <td>2009-12-10</td>\n",
       "      <td>64.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_code                         state patient_id  sledai visit_date  \\\n",
       "105  GSM3428415  Systemic Lupus Erythematosus       1001       4 2009-10-08   \n",
       "106  GSM3428416  Systemic Lupus Erythematosus       1001       2 2010-01-11   \n",
       "107  GSM3428417  Systemic Lupus Erythematosus       1001       6 2010-03-29   \n",
       "108  GSM3428418  Systemic Lupus Erythematosus       1041       2 2009-09-24   \n",
       "109  GSM3428419  Systemic Lupus Erythematosus       1041      10 2009-12-10   \n",
       "\n",
       "     neutrophil_percent  \n",
       "105                65.2  \n",
       "106                61.5  \n",
       "107                57.5  \n",
       "108                73.8  \n",
       "109                64.1  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "128f4770",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_code</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_title</th>\n",
       "      <th>gene_expression_value</th>\n",
       "      <th>main_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GSM3428310</td>\n",
       "      <td>1007_PM_s_at</td>\n",
       "      <td>DDR1</td>\n",
       "      <td>discoidin domain receptor tyrosine kinase 1</td>\n",
       "      <td>4.955257</td>\n",
       "      <td>DDR1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GSM3428310</td>\n",
       "      <td>1053_PM_at</td>\n",
       "      <td>RFC2</td>\n",
       "      <td>replication factor C (activator 1) 2, 40kDa</td>\n",
       "      <td>5.984784</td>\n",
       "      <td>RFC2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GSM3428310</td>\n",
       "      <td>117_PM_at</td>\n",
       "      <td>HSPA6</td>\n",
       "      <td>heat shock 70kDa protein 6 (HSP70B')</td>\n",
       "      <td>9.477945</td>\n",
       "      <td>HSPA6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GSM3428310</td>\n",
       "      <td>121_PM_at</td>\n",
       "      <td>PAX8</td>\n",
       "      <td>paired box 8</td>\n",
       "      <td>4.553229</td>\n",
       "      <td>PAX8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GSM3428310</td>\n",
       "      <td>1255_PM_g_at</td>\n",
       "      <td>GUCA1A</td>\n",
       "      <td>guanylate cyclase activator 1A (retina)</td>\n",
       "      <td>1.921190</td>\n",
       "      <td>GUCA1A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_code            ID gene_symbol  \\\n",
       "0  GSM3428310  1007_PM_s_at        DDR1   \n",
       "1  GSM3428310    1053_PM_at        RFC2   \n",
       "2  GSM3428310     117_PM_at       HSPA6   \n",
       "3  GSM3428310     121_PM_at        PAX8   \n",
       "4  GSM3428310  1255_PM_g_at      GUCA1A   \n",
       "\n",
       "                                    gene_title  gene_expression_value  \\\n",
       "0  discoidin domain receptor tyrosine kinase 1               4.955257   \n",
       "1  replication factor C (activator 1) 2, 40kDa               5.984784   \n",
       "2         heat shock 70kDa protein 6 (HSP70B')               9.477945   \n",
       "3                                 paired box 8               4.553229   \n",
       "4      guanylate cyclase activator 1A (retina)               1.921190   \n",
       "\n",
       "  main_gene  \n",
       "0      DDR1  \n",
       "1      RFC2  \n",
       "2     HSPA6  \n",
       "3      PAX8  \n",
       "4    GUCA1A  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56ded0b4",
   "metadata": {},
   "source": [
    "### Classificando lesões"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e68869e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_lesion(sledai_score):\n",
    "    if sledai_score == 0:\n",
    "        return 'healthy'\n",
    "    if 1 <= sledai_score <= 6:\n",
    "        return 'leve'\n",
    "    elif 7 <= sledai_score <= 11:\n",
    "        return 'media'\n",
    "    else:\n",
    "        return 'grave'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5ab4acff",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_data['class'] = patient_data['sledai'].apply(classify_lesion)\n",
    "patient_data['class_group'] = np.where(patient_data['class'].isin(['grave', 'media']), 'medio/grave', 'leve')\n",
    "patient_data['class_group'] = np.where(patient_data['class'] == 'healthy', 'healthy', patient_data['class_group'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ab6919dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_code</th>\n",
       "      <th>state</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sledai</th>\n",
       "      <th>visit_date</th>\n",
       "      <th>neutrophil_percent</th>\n",
       "      <th>class</th>\n",
       "      <th>class_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>GSM3428415</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>4</td>\n",
       "      <td>2009-10-08</td>\n",
       "      <td>65.2</td>\n",
       "      <td>leve</td>\n",
       "      <td>leve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>GSM3428416</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>2</td>\n",
       "      <td>2010-01-11</td>\n",
       "      <td>61.5</td>\n",
       "      <td>leve</td>\n",
       "      <td>leve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>GSM3428417</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>6</td>\n",
       "      <td>2010-03-29</td>\n",
       "      <td>57.5</td>\n",
       "      <td>leve</td>\n",
       "      <td>leve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>GSM3428418</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1041</td>\n",
       "      <td>2</td>\n",
       "      <td>2009-09-24</td>\n",
       "      <td>73.8</td>\n",
       "      <td>leve</td>\n",
       "      <td>leve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>GSM3428419</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1041</td>\n",
       "      <td>10</td>\n",
       "      <td>2009-12-10</td>\n",
       "      <td>64.1</td>\n",
       "      <td>media</td>\n",
       "      <td>medio/grave</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_code                         state patient_id  sledai visit_date  \\\n",
       "105  GSM3428415  Systemic Lupus Erythematosus       1001       4 2009-10-08   \n",
       "106  GSM3428416  Systemic Lupus Erythematosus       1001       2 2010-01-11   \n",
       "107  GSM3428417  Systemic Lupus Erythematosus       1001       6 2010-03-29   \n",
       "108  GSM3428418  Systemic Lupus Erythematosus       1041       2 2009-09-24   \n",
       "109  GSM3428419  Systemic Lupus Erythematosus       1041      10 2009-12-10   \n",
       "\n",
       "     neutrophil_percent  class  class_group  \n",
       "105                65.2   leve         leve  \n",
       "106                61.5   leve         leve  \n",
       "107                57.5   leve         leve  \n",
       "108                73.8   leve         leve  \n",
       "109                64.1  media  medio/grave  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1e148615",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data.to_parquet('../../data/processed/gene_data.parquet', index=False)\n",
    "patient_data.to_csv('../../data/processed/patient_data.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "d0cba9cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_code</th>\n",
       "      <th>state</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>sledai</th>\n",
       "      <th>visit_date</th>\n",
       "      <th>neutrophil_percent</th>\n",
       "      <th>class</th>\n",
       "      <th>class_group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>GSM3428415</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>4</td>\n",
       "      <td>2009-10-08</td>\n",
       "      <td>65.2</td>\n",
       "      <td>leve</td>\n",
       "      <td>leve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>GSM3428416</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>2</td>\n",
       "      <td>2010-01-11</td>\n",
       "      <td>61.5</td>\n",
       "      <td>leve</td>\n",
       "      <td>leve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>GSM3428417</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1001</td>\n",
       "      <td>6</td>\n",
       "      <td>2010-03-29</td>\n",
       "      <td>57.5</td>\n",
       "      <td>leve</td>\n",
       "      <td>leve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>GSM3428418</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1041</td>\n",
       "      <td>2</td>\n",
       "      <td>2009-09-24</td>\n",
       "      <td>73.8</td>\n",
       "      <td>leve</td>\n",
       "      <td>leve</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>GSM3428419</td>\n",
       "      <td>Systemic Lupus Erythematosus</td>\n",
       "      <td>1041</td>\n",
       "      <td>10</td>\n",
       "      <td>2009-12-10</td>\n",
       "      <td>64.1</td>\n",
       "      <td>media</td>\n",
       "      <td>medio/grave</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GSM3428325</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>healthy</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GSM3428326</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>healthy</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GSM3428327</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>healthy</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GSM3428328</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>healthy</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GSM3428329</td>\n",
       "      <td>Healthy</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>NaT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>healthy</td>\n",
       "      <td>healthy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>312 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_code                         state patient_id  sledai visit_date  \\\n",
       "105  GSM3428415  Systemic Lupus Erythematosus       1001       4 2009-10-08   \n",
       "106  GSM3428416  Systemic Lupus Erythematosus       1001       2 2010-01-11   \n",
       "107  GSM3428417  Systemic Lupus Erythematosus       1001       6 2010-03-29   \n",
       "108  GSM3428418  Systemic Lupus Erythematosus       1041       2 2009-09-24   \n",
       "109  GSM3428419  Systemic Lupus Erythematosus       1041      10 2009-12-10   \n",
       "..          ...                           ...        ...     ...        ...   \n",
       "15   GSM3428325                       Healthy        NaN       0        NaT   \n",
       "16   GSM3428326                       Healthy        NaN       0        NaT   \n",
       "17   GSM3428327                       Healthy        NaN       0        NaT   \n",
       "18   GSM3428328                       Healthy        NaN       0        NaT   \n",
       "19   GSM3428329                       Healthy        NaN       0        NaT   \n",
       "\n",
       "     neutrophil_percent    class  class_group  \n",
       "105                65.2     leve         leve  \n",
       "106                61.5     leve         leve  \n",
       "107                57.5     leve         leve  \n",
       "108                73.8     leve         leve  \n",
       "109                64.1    media  medio/grave  \n",
       "..                  ...      ...          ...  \n",
       "15                  NaN  healthy      healthy  \n",
       "16                  NaN  healthy      healthy  \n",
       "17                  NaN  healthy      healthy  \n",
       "18                  NaN  healthy      healthy  \n",
       "19                  NaN  healthy      healthy  \n",
       "\n",
       "[312 rows x 8 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4060e51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "hole": 0.34,
         "insidetextorientation": "radial",
         "labels": [
          "Grave",
          "Healthy",
          "Leve",
          "Media"
         ],
         "marker": {
          "colors": [
           "#ABACB2",
           "#bb77ba",
           "#39b3c2",
           "#7e7ae7"
          ]
         },
         "textfont": {
          "family": "Arial Black"
         },
         "textinfo": "percent+label",
         "type": "pie",
         "values": {
          "bdata": "FDuxEzuxgz9cvuVbvuXbP57YiZ3Yid0/uHzLt3zLtz8=",
          "dtype": "f8"
         }
        }
       ],
       "layout": {
        "showlegend": false,
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "white",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "white",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "#C8D4E3",
             "linecolor": "#C8D4E3",
             "minorgridcolor": "#C8D4E3",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "#C8D4E3",
             "linecolor": "#C8D4E3",
             "minorgridcolor": "#C8D4E3",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermap": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermap"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "white",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "#C8D4E3"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "white",
          "polar": {
           "angularaxis": {
            "gridcolor": "#EBF0F8",
            "linecolor": "#EBF0F8",
            "ticks": ""
           },
           "bgcolor": "white",
           "radialaxis": {
            "gridcolor": "#EBF0F8",
            "linecolor": "#EBF0F8",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "white",
            "gridcolor": "#DFE8F3",
            "gridwidth": 2,
            "linecolor": "#EBF0F8",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "#EBF0F8"
           },
           "yaxis": {
            "backgroundcolor": "white",
            "gridcolor": "#DFE8F3",
            "gridwidth": 2,
            "linecolor": "#EBF0F8",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "#EBF0F8"
           },
           "zaxis": {
            "backgroundcolor": "white",
            "gridcolor": "#DFE8F3",
            "gridwidth": 2,
            "linecolor": "#EBF0F8",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "#EBF0F8"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "#DFE8F3",
            "linecolor": "#A2B1C6",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "#DFE8F3",
            "linecolor": "#A2B1C6",
            "ticks": ""
           },
           "bgcolor": "white",
           "caxis": {
            "gridcolor": "#DFE8F3",
            "linecolor": "#A2B1C6",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "#EBF0F8",
           "linecolor": "#EBF0F8",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "#EBF0F8",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "#EBF0F8",
           "linecolor": "#EBF0F8",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "#EBF0F8",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "Distribuição do Grau de Lesão (SLEDAI) nos Pacientes",
         "x": 0.5
        },
        "width": 500
       }
      }
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "hole": 0.34,
         "insidetextorientation": "radial",
         "labels": [
          "Leve",
          "Medio/grave"
         ],
         "marker": {
          "colors": [
           "#39b3c2",
           "#7e7ae7"
          ]
         },
         "textfont": {
          "family": "Arial Black"
         },
         "textinfo": "percent+label",
         "type": "pie",
         "values": {
          "bdata": "L7rooosu6j9GF1100UXHPw==",
          "dtype": "f8"
         }
        }
       ],
       "layout": {
        "showlegend": false,
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "white",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "white",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "#C8D4E3",
             "linecolor": "#C8D4E3",
             "minorgridcolor": "#C8D4E3",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "#C8D4E3",
             "linecolor": "#C8D4E3",
             "minorgridcolor": "#C8D4E3",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermap": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermap"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "white",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "#C8D4E3"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "white",
          "polar": {
           "angularaxis": {
            "gridcolor": "#EBF0F8",
            "linecolor": "#EBF0F8",
            "ticks": ""
           },
           "bgcolor": "white",
           "radialaxis": {
            "gridcolor": "#EBF0F8",
            "linecolor": "#EBF0F8",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "white",
            "gridcolor": "#DFE8F3",
            "gridwidth": 2,
            "linecolor": "#EBF0F8",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "#EBF0F8"
           },
           "yaxis": {
            "backgroundcolor": "white",
            "gridcolor": "#DFE8F3",
            "gridwidth": 2,
            "linecolor": "#EBF0F8",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "#EBF0F8"
           },
           "zaxis": {
            "backgroundcolor": "white",
            "gridcolor": "#DFE8F3",
            "gridwidth": 2,
            "linecolor": "#EBF0F8",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "#EBF0F8"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "#DFE8F3",
            "linecolor": "#A2B1C6",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "#DFE8F3",
            "linecolor": "#A2B1C6",
            "ticks": ""
           },
           "bgcolor": "white",
           "caxis": {
            "gridcolor": "#DFE8F3",
            "linecolor": "#A2B1C6",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "#EBF0F8",
           "linecolor": "#EBF0F8",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "#EBF0F8",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "#EBF0F8",
           "linecolor": "#EBF0F8",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "#EBF0F8",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "Distribuição do Grau de Lesão (SLEDAI) nos Pacientes",
         "x": 0.5
        },
        "width": 500
       }
      }
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def plot_pie_distribution(patient_data, col='class'):\n",
    "    if col == 'class':\n",
    "        colors = [\"#ABACB2\", \"#bb77ba\", \"#39b3c2\", \"#7e7ae7\"]\n",
    "    else:\n",
    "        colors = [\"#39b3c2\", \"#7e7ae7\"]\n",
    "        \n",
    "    fig = go.Figure(data=[go.Pie(\n",
    "        labels=[label.capitalize() for label in np.sort(patient_data[col].unique())],\n",
    "        values=patient_data[col].value_counts(normalize=True).sort_index().values,\n",
    "        hole=0.34,\n",
    "        marker_colors=colors,\n",
    "        textinfo='percent+label',\n",
    "        insidetextorientation='radial',\n",
    "        textfont=dict(family='Arial Black')\n",
    "    )])\n",
    "    \n",
    "    fig.update_layout(\n",
    "        title=dict(x=0.5, text='Distribuição do Grau de Lesão (SLEDAI) nos Pacientes'),\n",
    "        template='plotly_white',\n",
    "        showlegend=False,\n",
    "        width=500\n",
    "    )\n",
    "    \n",
    "    fig.show()\n",
    "\n",
    "fig = plot_pie_distribution(patient_data, col='class')\n",
    "fig = plot_pie_distribution(patient_data.query(\"class_group != 'healthy'\"), col='class_group')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "94fd9fa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtra a observação com maior SLEDAI por paciente\n",
    "patient_data['max_sledai'] = patient_data.groupby('patient_id')['sledai'].transform('max')\n",
    "patient_data_max_sledai = patient_data.query(\"sledai == max_sledai\").drop(columns=['max_sledai'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ead3a952",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = gene_data.merge(patient_data_max_sledai, on='sample_code')\n",
    "# df = df[['sample_code', 'patient_id', 'state', 'sledai', 'visit_date', 'neutrophil_percent', 'gene_symbol', 'gene_title', 'gene_expression_value', 'ID_REF']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3a131e6c",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[26]\u001b[39m\u001b[32m, line 10\u001b[39m\n\u001b[32m      7\u001b[39m results = []\n\u001b[32m      9\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m gene \u001b[38;5;129;01min\u001b[39;00m genes:\n\u001b[32m---> \u001b[39m\u001b[32m10\u001b[39m     group1 = df[(df[\u001b[33m'\u001b[39m\u001b[33mmain_gene\u001b[39m\u001b[33m'\u001b[39m] == gene) & (\u001b[43mdf\u001b[49m\u001b[43m[\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mclass_group\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[43m==\u001b[49m\u001b[43m \u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mleve\u001b[39;49m\u001b[33;43m'\u001b[39;49m)][\u001b[33m'\u001b[39m\u001b[33mgene_expression_value\u001b[39m\u001b[33m'\u001b[39m]\n\u001b[32m     11\u001b[39m     group2 = df[(df[\u001b[33m'\u001b[39m\u001b[33mmain_gene\u001b[39m\u001b[33m'\u001b[39m] == gene) & (df[\u001b[33m'\u001b[39m\u001b[33mclass_group\u001b[39m\u001b[33m'\u001b[39m] == \u001b[33m'\u001b[39m\u001b[33mmedio/grave\u001b[39m\u001b[33m'\u001b[39m)][\u001b[33m'\u001b[39m\u001b[33mgene_expression_value\u001b[39m\u001b[33m'\u001b[39m]\n\u001b[32m     12\u001b[39m     stat, pval = stats.ttest_ind(group1, group2, equal_var=\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/ml/lib/python3.13/site-packages/pandas/core/ops/common.py:76\u001b[39m, in \u001b[36m_unpack_zerodim_and_defer.<locals>.new_method\u001b[39m\u001b[34m(self, other)\u001b[39m\n\u001b[32m     72\u001b[39m             \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mNotImplemented\u001b[39m\n\u001b[32m     74\u001b[39m other = item_from_zerodim(other)\n\u001b[32m---> \u001b[39m\u001b[32m76\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmethod\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mother\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/ml/lib/python3.13/site-packages/pandas/core/arraylike.py:40\u001b[39m, in \u001b[36mOpsMixin.__eq__\u001b[39m\u001b[34m(self, other)\u001b[39m\n\u001b[32m     38\u001b[39m \u001b[38;5;129m@unpack_zerodim_and_defer\u001b[39m(\u001b[33m\"\u001b[39m\u001b[33m__eq__\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m     39\u001b[39m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[34m__eq__\u001b[39m(\u001b[38;5;28mself\u001b[39m, other):\n\u001b[32m---> \u001b[39m\u001b[32m40\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_cmp_method\u001b[49m\u001b[43m(\u001b[49m\u001b[43mother\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43moperator\u001b[49m\u001b[43m.\u001b[49m\u001b[43meq\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/ml/lib/python3.13/site-packages/pandas/core/series.py:6138\u001b[39m, in \u001b[36mSeries._cmp_method\u001b[39m\u001b[34m(self, other, op)\u001b[39m\n\u001b[32m   6135\u001b[39m lvalues = \u001b[38;5;28mself\u001b[39m._values\n\u001b[32m   6136\u001b[39m rvalues = extract_array(other, extract_numpy=\u001b[38;5;28;01mTrue\u001b[39;00m, extract_range=\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[32m-> \u001b[39m\u001b[32m6138\u001b[39m res_values = \u001b[43mops\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcomparison_op\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m   6140\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m._construct_result(res_values, name=res_name)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/ml/lib/python3.13/site-packages/pandas/core/ops/array_ops.py:344\u001b[39m, in \u001b[36mcomparison_op\u001b[39m\u001b[34m(left, right, op)\u001b[39m\n\u001b[32m    341\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m invalid_comparison(lvalues, rvalues, op)\n\u001b[32m    343\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m lvalues.dtype == \u001b[38;5;28mobject\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(rvalues, \u001b[38;5;28mstr\u001b[39m):\n\u001b[32m--> \u001b[39m\u001b[32m344\u001b[39m     res_values = \u001b[43mcomp_method_OBJECT_ARRAY\u001b[49m\u001b[43m(\u001b[49m\u001b[43mop\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlvalues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrvalues\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    346\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m    347\u001b[39m     res_values = _na_arithmetic_op(lvalues, rvalues, op, is_cmp=\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/anaconda3/envs/ml/lib/python3.13/site-packages/pandas/core/ops/array_ops.py:129\u001b[39m, in \u001b[36mcomp_method_OBJECT_ARRAY\u001b[39m\u001b[34m(op, x, y)\u001b[39m\n\u001b[32m    127\u001b[39m     result = libops.vec_compare(x.ravel(), y.ravel(), op)\n\u001b[32m    128\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m129\u001b[39m     result = \u001b[43mlibops\u001b[49m\u001b[43m.\u001b[49m\u001b[43mscalar_compare\u001b[49m\u001b[43m(\u001b[49m\u001b[43mx\u001b[49m\u001b[43m.\u001b[49m\u001b[43mravel\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43my\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mop\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    130\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m result.reshape(x.shape)\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "# Supondo que df tem colunas: 'gene_symbol', 'gene_expression_value', 'class'\n",
    "# Exemplo: comparar 'leve' vs 'grave'\n",
    "\n",
    "genes = df['main_gene'].dropna().unique()\n",
    "results = []\n",
    "\n",
    "for gene in genes:\n",
    "    group1 = df[(df['main_gene'] == gene) & (df['class_group'] == 'leve')]['gene_expression_value']\n",
    "    group2 = df[(df['main_gene'] == gene) & (df['class_group'] == 'medio/grave')]['gene_expression_value']\n",
    "    stat, pval = stats.ttest_ind(group1, group2, equal_var=False)\n",
    "    results.append({'gene': gene, 'pval': pval, 'mean_leve': group1.mean(), 'mean_grave': group2.mean()})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cfc5745",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DataFrame com resultados\n",
    "res_df = pd.DataFrame(results)\n",
    "res_df['log2fc'] = np.log2(res_df['mean_grave'] + 1) - np.log2(res_df['mean_leve'] + 1)\n",
    "res_df = res_df.sort_values('pval')\n",
    "\n",
    "# Genes mais diferenciais\n",
    "top_genes = res_df[res_df['pval'] < 0.05].sort_values('log2fc', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b3e4c4a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'gene': 'DDR1',\n",
       "  'pval': np.float64(0.2100797001084677),\n",
       "  'mean_leve': np.float64(4.425244505479166),\n",
       "  'mean_grave': np.float64(4.369157444695312)},\n",
       " {'gene': 'RFC2',\n",
       "  'pval': np.float64(0.20727321476765737),\n",
       "  'mean_leve': np.float64(6.259382690402777),\n",
       "  'mean_grave': np.float64(6.191040201328125)},\n",
       " {'gene': 'HSPA6',\n",
       "  'pval': np.float64(0.9562692446052213),\n",
       "  'mean_leve': np.float64(9.587159257618055),\n",
       "  'mean_grave': np.float64(9.582517183390625)},\n",
       " {'gene': 'PAX8',\n",
       "  'pval': np.float64(0.9477600426952131),\n",
       "  'mean_leve': np.float64(3.476633524646701),\n",
       "  'mean_grave': np.float64(3.4725149381523437)},\n",
       " {'gene': 'GUCA1A',\n",
       "  'pval': np.float64(0.8995143302419972),\n",
       "  'mean_leve': np.float64(3.27648892130787),\n",
       "  'mean_grave': np.float64(3.2895532785624995)},\n",
       " {'gene': 'UBA7',\n",
       "  'pval': np.float64(0.027651728205105915),\n",
       "  'mean_leve': np.float64(7.942164412815973),\n",
       "  'mean_grave': np.float64(7.825528932)},\n",
       " {'gene': 'THRA',\n",
       "  'pval': np.float64(0.9894981719394134),\n",
       "  'mean_leve': np.float64(4.843992193317708),\n",
       "  'mean_grave': np.float64(4.844992044609375)},\n",
       " {'gene': 'PTPN21',\n",
       "  'pval': np.float64(0.9269691282758964),\n",
       "  'mean_leve': np.float64(2.9486750925736107),\n",
       "  'mean_grave': np.float64(2.95615143126875)},\n",
       " {'gene': 'CCL5',\n",
       "  'pval': np.float64(0.9185564549795817),\n",
       "  'mean_leve': np.float64(8.86170368886111),\n",
       "  'mean_grave': np.float64(8.851775397979166)},\n",
       " {'gene': 'CYP2E1',\n",
       "  'pval': np.float64(0.8166954106986392),\n",
       "  'mean_leve': np.float64(3.5301438635572917),\n",
       "  'mean_grave': np.float64(3.5112960784140625)},\n",
       " {'gene': 'EPHB3',\n",
       "  'pval': np.float64(0.9423019184379623),\n",
       "  'mean_leve': np.float64(4.372970666454861),\n",
       "  'mean_grave': np.float64(4.383114266625)},\n",
       " {'gene': 'ESRRA',\n",
       "  'pval': np.float64(0.5667974747105806),\n",
       "  'mean_leve': np.float64(5.81350716771875),\n",
       "  'mean_grave': np.float64(5.768896700359375)},\n",
       " {'gene': 'CYP2A6',\n",
       "  'pval': np.float64(0.7127334378281333),\n",
       "  'mean_leve': np.float64(3.50590681528125),\n",
       "  'mean_grave': np.float64(3.5265691974140623)},\n",
       " {'gene': 'SCARB1',\n",
       "  'pval': np.float64(0.40141698835486805),\n",
       "  'mean_leve': np.float64(3.779313699769097),\n",
       "  'mean_grave': np.float64(3.828603049703125)},\n",
       " {'gene': 'TTLL12',\n",
       "  'pval': np.float64(0.06577243177583803),\n",
       "  'mean_leve': np.float64(6.577981674958333),\n",
       "  'mean_grave': np.float64(6.481751999546875)},\n",
       " {'gene': 'NCRNA00152',\n",
       "  'pval': np.float64(0.17736400047135262),\n",
       "  'mean_leve': np.float64(5.1813174451875),\n",
       "  'mean_grave': np.float64(5.04211678709375)},\n",
       " {'gene': 'WFDC2',\n",
       "  'pval': np.float64(0.9490725041347027),\n",
       "  'mean_leve': np.float64(4.023718000233796),\n",
       "  'mean_grave': np.float64(4.0277645161562505)},\n",
       " {'gene': 'MAPK1',\n",
       "  'pval': np.float64(0.829876377317297),\n",
       "  'mean_leve': np.float64(7.220351087847222),\n",
       "  'mean_grave': np.float64(7.1975782686205365)},\n",
       " {'gene': 'ADAM32',\n",
       "  'pval': np.float64(0.29531787861113773),\n",
       "  'mean_leve': np.float64(2.7517382842499996),\n",
       "  'mean_grave': np.float64(2.8108132890000004)},\n",
       " {'gene': 'SPATA17',\n",
       "  'pval': np.float64(0.6650762529005492),\n",
       "  'mean_leve': np.float64(1.8808886966111114),\n",
       "  'mean_grave': np.float64(1.9045919819687498)},\n",
       " {'gene': 'PRR22',\n",
       "  'pval': np.float64(0.7688571504812888),\n",
       "  'mean_leve': np.float64(4.132249850833333),\n",
       "  'mean_grave': np.float64(4.160537322031249)},\n",
       " {'gene': 'PXK',\n",
       "  'pval': np.float64(0.9085806610403755),\n",
       "  'mean_leve': np.float64(5.905887383518518),\n",
       "  'mean_grave': np.float64(5.893004812458334)},\n",
       " {'gene': 'VPS18',\n",
       "  'pval': np.float64(0.7572325735113566),\n",
       "  'mean_leve': np.float64(4.863292538052083),\n",
       "  'mean_grave': np.float64(4.920251463406251)},\n",
       " {'gene': 'C9orf30',\n",
       "  'pval': np.float64(0.3373326442383505),\n",
       "  'mean_leve': np.float64(6.3774254423020835),\n",
       "  'mean_grave': np.float64(6.128499743171875)},\n",
       " {'gene': 'SLC46A1',\n",
       "  'pval': np.float64(0.8435688069234317),\n",
       "  'mean_leve': np.float64(3.446652223668403),\n",
       "  'mean_grave': np.float64(3.4608826533203123)},\n",
       " {'gene': 'TIMD4',\n",
       "  'pval': np.float64(0.6942945003270335),\n",
       "  'mean_leve': np.float64(3.203758291125),\n",
       "  'mean_grave': np.float64(3.2408571683125)},\n",
       " {'gene': 'SLC39A5',\n",
       "  'pval': np.float64(0.5004846814692866),\n",
       "  'mean_leve': np.float64(4.379377232972222),\n",
       "  'mean_grave': np.float64(4.41331723703125)},\n",
       " {'gene': 'ZDHHC11',\n",
       "  'pval': np.float64(0.7124160605052827),\n",
       "  'mean_leve': np.float64(4.87848725036111),\n",
       "  'mean_grave': np.float64(4.928120550770833)},\n",
       " {'gene': 'ATP6V1E2',\n",
       "  'pval': np.float64(0.3841355584287157),\n",
       "  'mean_leve': np.float64(5.804001394979166),\n",
       "  'mean_grave': np.float64(5.74132397246875)},\n",
       " {'gene': 'AFG3L1',\n",
       "  'pval': np.float64(0.5440014273356237),\n",
       "  'mean_leve': np.float64(5.235680045319445),\n",
       "  'mean_grave': np.float64(5.184569603447916)},\n",
       " {'gene': 'CILP2',\n",
       "  'pval': np.float64(0.43366452608432593),\n",
       "  'mean_leve': np.float64(3.442805505770833),\n",
       "  'mean_grave': np.float64(3.53321888378125)},\n",
       " {'gene': 'PIGX',\n",
       "  'pval': np.float64(0.7646281261980106),\n",
       "  'mean_leve': np.float64(6.695964870138888),\n",
       "  'mean_grave': np.float64(6.723067920187501)},\n",
       " {'gene': 'TMEM196',\n",
       "  'pval': np.float64(0.6948106547084509),\n",
       "  'mean_leve': np.float64(1.9986814075902777),\n",
       "  'mean_grave': np.float64(1.96777766153125)},\n",
       " {'gene': 'SLC39A13',\n",
       "  'pval': np.float64(0.6860159792706966),\n",
       "  'mean_leve': np.float64(5.447710741791667),\n",
       "  'mean_grave': np.float64(5.38533620403125)},\n",
       " {'gene': 'BEST4',\n",
       "  'pval': np.float64(0.6497609562895582),\n",
       "  'mean_leve': np.float64(3.682348932861111),\n",
       "  'mean_grave': np.float64(3.65176512925)},\n",
       " {'gene': 'AKD1',\n",
       "  'pval': np.float64(0.8083552491000551),\n",
       "  'mean_leve': np.float64(4.429808533627315),\n",
       "  'mean_grave': np.float64(4.4001703119375)},\n",
       " {'gene': 'CORO6',\n",
       "  'pval': np.float64(0.43453521365778947),\n",
       "  'mean_leve': np.float64(3.3213862406180557),\n",
       "  'mean_grave': np.float64(3.3867495742812497)},\n",
       " {'gene': 'TMEM106A',\n",
       "  'pval': np.float64(0.899560391875013),\n",
       "  'mean_leve': np.float64(4.470841683508681),\n",
       "  'mean_grave': np.float64(4.4557740650468745)},\n",
       " {'gene': 'ALG10',\n",
       "  'pval': np.float64(0.47553461258919405),\n",
       "  'mean_leve': np.float64(2.943993335013889),\n",
       "  'mean_grave': np.float64(3.010002432265625)},\n",
       " {'gene': 'TTC39C',\n",
       "  'pval': np.float64(0.8679163163538434),\n",
       "  'mean_leve': np.float64(6.8174505917430555),\n",
       "  'mean_grave': np.float64(6.7889747008375)},\n",
       " {'gene': 'NEXN',\n",
       "  'pval': np.float64(0.38814735847976933),\n",
       "  'mean_leve': np.float64(6.333273162545138),\n",
       "  'mean_grave': np.float64(6.207026367921875)},\n",
       " {'gene': 'C15orf40',\n",
       "  'pval': np.float64(0.34126604052969034),\n",
       "  'mean_leve': np.float64(4.953533331395834),\n",
       "  'mean_grave': np.float64(4.8198577879687505)},\n",
       " {'gene': 'RAX2',\n",
       "  'pval': np.float64(0.9873548594925986),\n",
       "  'mean_leve': np.float64(6.038674044243056),\n",
       "  'mean_grave': np.float64(6.037621903)},\n",
       " {'gene': 'MFAP3',\n",
       "  'pval': np.float64(0.947761085403238),\n",
       "  'mean_leve': np.float64(5.228990595671297),\n",
       "  'mean_grave': np.float64(5.24221380665625)},\n",
       " {'gene': 'EYA3',\n",
       "  'pval': np.float64(0.9853848801984859),\n",
       "  'mean_leve': np.float64(4.611025046074652),\n",
       "  'mean_grave': np.float64(4.6071968528125)},\n",
       " {'gene': 'GIMAP1',\n",
       "  'pval': np.float64(0.8132285324942073),\n",
       "  'mean_leve': np.float64(6.027559787813889),\n",
       "  'mean_grave': np.float64(5.986055106524999)},\n",
       " {'gene': 'KLK8',\n",
       "  'pval': np.float64(0.6643389623032234),\n",
       "  'mean_leve': np.float64(2.9142139495729165),\n",
       "  'mean_grave': np.float64(2.951348084484375)},\n",
       " {'gene': 'CCDC65',\n",
       "  'pval': np.float64(0.9297633401315848),\n",
       "  'mean_leve': np.float64(4.507281521421296),\n",
       "  'mean_grave': np.float64(4.51979474559375)},\n",
       " {'gene': 'FAM122C',\n",
       "  'pval': np.float64(0.889766100983463),\n",
       "  'mean_leve': np.float64(3.395734533890625),\n",
       "  'mean_grave': np.float64(3.411973659132813)},\n",
       " {'gene': 'CCDC11',\n",
       "  'pval': np.float64(0.26352518436001454),\n",
       "  'mean_leve': np.float64(2.8954075277604168),\n",
       "  'mean_grave': np.float64(2.9839654745781248)},\n",
       " {'gene': 'ARMCX4',\n",
       "  'pval': np.float64(0.5512952883061364),\n",
       "  'mean_leve': np.float64(2.497729853777778),\n",
       "  'mean_grave': np.float64(2.45672219969375)},\n",
       " {'gene': 'RBBP6',\n",
       "  'pval': np.float64(0.9345778168007292),\n",
       "  'mean_leve': np.float64(6.112858435688368),\n",
       "  'mean_grave': np.float64(6.120139847105468)},\n",
       " {'gene': 'CENPBD1',\n",
       "  'pval': np.float64(0.31765020327538995),\n",
       "  'mean_leve': np.float64(4.669672651475695),\n",
       "  'mean_grave': np.float64(4.5136517403125005)},\n",
       " {'gene': 'TRIOBP',\n",
       "  'pval': np.float64(0.8515297413495396),\n",
       "  'mean_leve': np.float64(6.855529634484721),\n",
       "  'mean_grave': np.float64(6.8740362532312504)},\n",
       " {'gene': 'CATSPER1',\n",
       "  'pval': np.float64(0.7204358416888333),\n",
       "  'mean_leve': np.float64(3.8243003692777773),\n",
       "  'mean_grave': np.float64(3.7981456755625)},\n",
       " {'gene': 'HOXD4',\n",
       "  'pval': np.float64(0.042631316110911956),\n",
       "  'mean_leve': np.float64(4.291920129173611),\n",
       "  'mean_grave': np.float64(4.4112352188125)},\n",
       " {'gene': 'GSC',\n",
       "  'pval': np.float64(0.878612076960814),\n",
       "  'mean_leve': np.float64(2.3754763248055557),\n",
       "  'mean_grave': np.float64(2.38391775428125)},\n",
       " {'gene': 'SP7',\n",
       "  'pval': np.float64(0.5586178159204533),\n",
       "  'mean_leve': np.float64(2.2381258563750004),\n",
       "  'mean_grave': np.float64(2.2040941495937503)},\n",
       " {'gene': 'PDE7A',\n",
       "  'pval': np.float64(0.9091719560171995),\n",
       "  'mean_leve': np.float64(6.355969359352778),\n",
       "  'mean_grave': np.float64(6.3352697206875)},\n",
       " {'gene': 'CNOT7',\n",
       "  'pval': np.float64(0.9961207403729733),\n",
       "  'mean_leve': np.float64(6.832079137949652),\n",
       "  'mean_grave': np.float64(6.833387292382812)},\n",
       " {'gene': 'CRYZL1',\n",
       "  'pval': np.float64(0.8549584015940375),\n",
       "  'mean_leve': np.float64(5.216327427806944),\n",
       "  'mean_grave': np.float64(5.18331642145625)},\n",
       " {'gene': 'PRSS33',\n",
       "  'pval': np.float64(0.061007936169440335),\n",
       "  'mean_leve': np.float64(6.6486110872118065),\n",
       "  'mean_grave': np.float64(6.291712445296875)},\n",
       " {'gene': 'C19orf26',\n",
       "  'pval': np.float64(0.852216823837547),\n",
       "  'mean_leve': np.float64(3.6485187947361113),\n",
       "  'mean_grave': np.float64(3.6284773610624996)},\n",
       " {'gene': 'C8orf45',\n",
       "  'pval': np.float64(0.905066792314325),\n",
       "  'mean_leve': np.float64(2.1118424043854165),\n",
       "  'mean_grave': np.float64(2.11666203525)},\n",
       " {'gene': 'TIRAP',\n",
       "  'pval': np.float64(0.6776932558459702),\n",
       "  'mean_leve': np.float64(4.993709837824652),\n",
       "  'mean_grave': np.float64(4.934506053523437)},\n",
       " {'gene': 'LEAP2',\n",
       "  'pval': np.float64(0.5359485639876471),\n",
       "  'mean_leve': np.float64(4.564851750972222),\n",
       "  'mean_grave': np.float64(4.509995976375)},\n",
       " {'gene': 'MSI2',\n",
       "  'pval': np.float64(0.4408136439789049),\n",
       "  'mean_leve': np.float64(4.782552141938368),\n",
       "  'mean_grave': np.float64(4.691704070980469)},\n",
       " {'gene': 'SCIN',\n",
       "  'pval': np.float64(0.953905521228953),\n",
       "  'mean_leve': np.float64(3.362528907152778),\n",
       "  'mean_grave': np.float64(3.3489769449479163)},\n",
       " {'gene': 'CTCFL',\n",
       "  'pval': np.float64(0.312183319110918),\n",
       "  'mean_leve': np.float64(2.196336535923611),\n",
       "  'mean_grave': np.float64(2.2736979234375)},\n",
       " {'gene': 'C4orf33',\n",
       "  'pval': np.float64(0.8929087622815425),\n",
       "  'mean_leve': np.float64(3.3048364192129633),\n",
       "  'mean_grave': np.float64(3.330762848229167)},\n",
       " {'gene': 'ZNF333',\n",
       "  'pval': np.float64(0.8720703902534246),\n",
       "  'mean_leve': np.float64(4.897438448156249),\n",
       "  'mean_grave': np.float64(4.909114404755209)},\n",
       " {'gene': 'FAM18B2',\n",
       "  'pval': np.float64(0.08571208171345661),\n",
       "  'mean_leve': np.float64(4.731052639423611),\n",
       "  'mean_grave': np.float64(4.624875107625)},\n",
       " {'gene': 'RDH10',\n",
       "  'pval': np.float64(0.6582257376650493),\n",
       "  'mean_leve': np.float64(3.6631571793385422),\n",
       "  'mean_grave': np.float64(3.7255583859062504)},\n",
       " {'gene': 'SFRS13B',\n",
       "  'pval': np.float64(0.7966786446909926),\n",
       "  'mean_leve': np.float64(2.906494607479167),\n",
       "  'mean_grave': np.float64(2.887067184140625)},\n",
       " {'gene': 'FAM71A',\n",
       "  'pval': np.float64(0.9455776509156446),\n",
       "  'mean_leve': np.float64(4.387740443717592),\n",
       "  'mean_grave': np.float64(4.396256590385417)},\n",
       " {'gene': 'GAPT',\n",
       "  'pval': np.float64(0.39056886762169574),\n",
       "  'mean_leve': np.float64(6.499513743368055),\n",
       "  'mean_grave': np.float64(6.59056209434375)},\n",
       " {'gene': 'FLJ30901',\n",
       "  'pval': np.float64(0.8418080276257941),\n",
       "  'mean_leve': np.float64(4.2052773049236105),\n",
       "  'mean_grave': np.float64(4.1905422628124995)},\n",
       " {'gene': 'C8orf47',\n",
       "  'pval': np.float64(0.46948390643083304),\n",
       "  'mean_leve': np.float64(1.7680929723958334),\n",
       "  'mean_grave': np.float64(1.73746285953125)},\n",
       " {'gene': 'C1orf65',\n",
       "  'pval': np.float64(0.295231851983009),\n",
       "  'mean_leve': np.float64(4.814685369659723),\n",
       "  'mean_grave': np.float64(4.7733690439375)},\n",
       " {'gene': 'ENTHD1',\n",
       "  'pval': np.float64(0.7689953901623339),\n",
       "  'mean_leve': np.float64(2.2904008904479167),\n",
       "  'mean_grave': np.float64(2.2646798188281254)},\n",
       " {'gene': 'TSSK3',\n",
       "  'pval': np.float64(0.4775716351205631),\n",
       "  'mean_leve': np.float64(3.807250416305555),\n",
       "  'mean_grave': np.float64(3.7602379453281247)},\n",
       " {'gene': 'WFDC6',\n",
       "  'pval': np.float64(0.2675808368936039),\n",
       "  'mean_leve': np.float64(2.9381762945069445),\n",
       "  'mean_grave': np.float64(2.875087926375)},\n",
       " {'gene': 'CLEC12A /// CLEC12B',\n",
       "  'pval': np.float64(0.019430254295766447),\n",
       "  'mean_leve': np.float64(9.58762858057639),\n",
       "  'mean_grave': np.float64(9.99505363121875)},\n",
       " {'gene': 'BRF1',\n",
       "  'pval': np.float64(0.8285908365616356),\n",
       "  'mean_leve': np.float64(4.3968428336493055),\n",
       "  'mean_grave': np.float64(4.377392559890625)},\n",
       " {'gene': 'C15orf27',\n",
       "  'pval': np.float64(0.3726844752290064),\n",
       "  'mean_leve': np.float64(2.8748907523055554),\n",
       "  'mean_grave': np.float64(2.817684678875)},\n",
       " {'gene': 'C21orf41',\n",
       "  'pval': np.float64(0.23940517833763095),\n",
       "  'mean_leve': np.float64(4.405924764701389),\n",
       "  'mean_grave': np.float64(4.343910560031249)},\n",
       " {'gene': 'CALML6',\n",
       "  'pval': np.float64(0.9008568160305825),\n",
       "  'mean_leve': np.float64(2.7297639611458333),\n",
       "  'mean_grave': np.float64(2.72140361609375)},\n",
       " {'gene': 'NLRP5',\n",
       "  'pval': np.float64(0.9170863944232546),\n",
       "  'mean_leve': np.float64(3.461785780465278),\n",
       "  'mean_grave': np.float64(3.46804704821875)},\n",
       " {'gene': 'ODF4',\n",
       "  'pval': np.float64(0.4291567557957351),\n",
       "  'mean_leve': np.float64(2.9525547249826385),\n",
       "  'mean_grave': np.float64(2.904317739640625)},\n",
       " {'gene': 'CLEC4F',\n",
       "  'pval': np.float64(0.40774723657345446),\n",
       "  'mean_leve': np.float64(4.670707096666667),\n",
       "  'mean_grave': np.float64(4.83483041534375)},\n",
       " {'gene': 'DEFB106A /// DEFB106B',\n",
       "  'pval': np.float64(0.8034352629311983),\n",
       "  'mean_leve': np.float64(2.2726535440381945),\n",
       "  'mean_grave': np.float64(2.24868128759375)},\n",
       " {'gene': 'WFDC9',\n",
       "  'pval': np.float64(0.90364946880086),\n",
       "  'mean_leve': np.float64(2.8111960228437503),\n",
       "  'mean_grave': np.float64(2.81846427478125)},\n",
       " {'gene': 'NEDD1',\n",
       "  'pval': np.float64(0.820212774216147),\n",
       "  'mean_leve': np.float64(5.833448435865741),\n",
       "  'mean_grave': np.float64(5.883074338427083)},\n",
       " {'gene': 'TTLL10',\n",
       "  'pval': np.float64(0.8811024088938286),\n",
       "  'mean_leve': np.float64(3.8378936541458333),\n",
       "  'mean_grave': np.float64(3.8449017180208336)},\n",
       " {'gene': 'CALR3',\n",
       "  'pval': np.float64(0.7383075988844829),\n",
       "  'mean_leve': np.float64(3.389800411388889),\n",
       "  'mean_grave': np.float64(3.4112517552500004)},\n",
       " {'gene': 'C10orf25',\n",
       "  'pval': np.float64(0.18586013783643615),\n",
       "  'mean_leve': np.float64(3.3051095316944443),\n",
       "  'mean_grave': np.float64(3.2061170814999995)},\n",
       " {'gene': 'ETV3',\n",
       "  'pval': np.float64(0.982977316564309),\n",
       "  'mean_leve': np.float64(5.406481203724537),\n",
       "  'mean_grave': np.float64(5.402785838864584)},\n",
       " {'gene': 'KLHL10',\n",
       "  'pval': np.float64(0.6115956796203971),\n",
       "  'mean_leve': np.float64(2.7871735503854165),\n",
       "  'mean_grave': np.float64(2.7619443725937503)},\n",
       " {'gene': 'TM2D3',\n",
       "  'pval': np.float64(0.8021970720127688),\n",
       "  'mean_leve': np.float64(9.808999563256943),\n",
       "  'mean_grave': np.float64(9.832516421390626)},\n",
       " {'gene': 'ZNF485',\n",
       "  'pval': np.float64(0.3814790185547111),\n",
       "  'mean_leve': np.float64(3.1505189419583335),\n",
       "  'mean_grave': np.float64(3.2034040712499996)},\n",
       " {'gene': 'WDR17',\n",
       "  'pval': np.float64(0.7081777596732773),\n",
       "  'mean_leve': np.float64(1.894651859503472),\n",
       "  'mean_grave': np.float64(1.923596155734375)},\n",
       " {'gene': 'MFSD6L',\n",
       "  'pval': np.float64(0.8191519906742764),\n",
       "  'mean_leve': np.float64(4.652260570375),\n",
       "  'mean_grave': np.float64(4.637244130125)},\n",
       " {'gene': 'CDH23',\n",
       "  'pval': np.float64(0.8873809242074),\n",
       "  'mean_leve': np.float64(4.7707522986788184),\n",
       "  'mean_grave': np.float64(4.779416874492188)},\n",
       " {'gene': 'ANKAR',\n",
       "  'pval': np.float64(0.8649277514058574),\n",
       "  'mean_leve': np.float64(2.103372853527778),\n",
       "  'mean_grave': np.float64(2.08656406496875)},\n",
       " {'gene': 'MEGF11',\n",
       "  'pval': np.float64(0.9821830784374136),\n",
       "  'mean_leve': np.float64(3.206554445886574),\n",
       "  'mean_grave': np.float64(3.2023422339791665)},\n",
       " {'gene': 'GPR182',\n",
       "  'pval': np.float64(0.8359046530801365),\n",
       "  'mean_leve': np.float64(3.9225054466840277),\n",
       "  'mean_grave': np.float64(3.932477241953125)},\n",
       " {'gene': 'ESX1',\n",
       "  'pval': np.float64(0.9962259621606904),\n",
       "  'mean_leve': np.float64(1.678319919347222),\n",
       "  'mean_grave': np.float64(1.678138944125)},\n",
       " {'gene': 'C8orf12',\n",
       "  'pval': np.float64(0.4385551063149238),\n",
       "  'mean_leve': np.float64(1.8522257162152778),\n",
       "  'mean_grave': np.float64(1.81076253884375)},\n",
       " {'gene': 'SCGB1C1',\n",
       "  'pval': np.float64(0.21971986445975544),\n",
       "  'mean_leve': np.float64(4.043817668479167),\n",
       "  'mean_grave': np.float64(3.88227968303125)},\n",
       " {'gene': 'DNAJC5G',\n",
       "  'pval': np.float64(0.33159463351699203),\n",
       "  'mean_leve': np.float64(2.1668210414374998),\n",
       "  'mean_grave': np.float64(2.1084253046875)},\n",
       " {'gene': 'WDR88',\n",
       "  'pval': np.float64(0.9056391278650716),\n",
       "  'mean_leve': np.float64(2.7606266010659724),\n",
       "  'mean_grave': np.float64(2.7684223885468753)},\n",
       " {'gene': 'PRUNE2',\n",
       "  'pval': np.float64(0.5688079004340094),\n",
       "  'mean_leve': np.float64(2.8346024156365734),\n",
       "  'mean_grave': np.float64(2.7717635945)},\n",
       " {'gene': 'MBD3L2',\n",
       "  'pval': np.float64(0.17639934207317706),\n",
       "  'mean_leve': np.float64(3.205546909388889),\n",
       "  'mean_grave': np.float64(3.1269819511875)},\n",
       " {'gene': 'ADAMTSL1',\n",
       "  'pval': np.float64(0.9079107007100828),\n",
       "  'mean_leve': np.float64(2.813359699246528),\n",
       "  'mean_grave': np.float64(2.8209961198072917)},\n",
       " {'gene': 'MBD3L1',\n",
       "  'pval': np.float64(0.9283491397610182),\n",
       "  'mean_leve': np.float64(3.8121008502152782),\n",
       "  'mean_grave': np.float64(3.82061372778125)},\n",
       " {'gene': 'FAM46D',\n",
       "  'pval': np.float64(0.2901138484293772),\n",
       "  'mean_leve': np.float64(1.955491729263889),\n",
       "  'mean_grave': np.float64(1.86582747259375)},\n",
       " {'gene': 'SERPINB11',\n",
       "  'pval': np.float64(0.5505431761271651),\n",
       "  'mean_leve': np.float64(1.9577501124236112),\n",
       "  'mean_grave': np.float64(1.97924461140625)},\n",
       " {'gene': 'NCRNA00161',\n",
       "  'pval': np.float64(0.2885538232471834),\n",
       "  'mean_leve': np.float64(2.0435406286388886),\n",
       "  'mean_grave': np.float64(2.09137635809375)},\n",
       " {'gene': 'DSCR10',\n",
       "  'pval': np.float64(0.8096983817733351),\n",
       "  'mean_leve': np.float64(3.4620389562291667),\n",
       "  'mean_grave': np.float64(3.4745144992812502)},\n",
       " {'gene': 'ABHD11',\n",
       "  'pval': np.float64(0.6334179567136871),\n",
       "  'mean_leve': np.float64(4.692690398872685),\n",
       "  'mean_grave': np.float64(4.62093313978125)},\n",
       " {'gene': 'ACAP2',\n",
       "  'pval': np.float64(0.19276728194301546),\n",
       "  'mean_leve': np.float64(8.686676895739584),\n",
       "  'mean_grave': np.float64(8.802636951937501)},\n",
       " {'gene': 'GAMT',\n",
       "  'pval': np.float64(0.9528390375128943),\n",
       "  'mean_leve': np.float64(3.4198549819305555),\n",
       "  'mean_grave': np.float64(3.4175286324062504)},\n",
       " {'gene': 'PLCD3',\n",
       "  'pval': np.float64(0.9684400952946639),\n",
       "  'mean_leve': np.float64(4.349864958298612),\n",
       "  'mean_grave': np.float64(4.3359742595781245)},\n",
       " {'gene': 'IRF6',\n",
       "  'pval': np.float64(0.8795488348027011),\n",
       "  'mean_leve': np.float64(3.066732486935185),\n",
       "  'mean_grave': np.float64(3.0810289184270836)},\n",
       " {'gene': 'PTPRC',\n",
       "  'pval': np.float64(0.7731304589125025),\n",
       "  'mean_leve': np.float64(8.713430932122222),\n",
       "  'mean_grave': np.float64(8.76982284215625)},\n",
       " {'gene': 'MAN1A2',\n",
       "  'pval': np.float64(0.8928346203927795),\n",
       "  'mean_leve': np.float64(4.094264618806944),\n",
       "  'mean_grave': np.float64(4.07556649269375)},\n",
       " {'gene': 'RAPH1',\n",
       "  'pval': np.float64(0.7367695091428673),\n",
       "  'mean_leve': np.float64(4.1092904252625),\n",
       "  'mean_grave': np.float64(4.1633105833125)},\n",
       " {'gene': 'SMCR8',\n",
       "  'pval': np.float64(0.9376939944944214),\n",
       "  'mean_leve': np.float64(5.419749808238889),\n",
       "  'mean_grave': np.float64(5.43397508733125)},\n",
       " {'gene': 'LACTB',\n",
       "  'pval': np.float64(0.6774044112993007),\n",
       "  'mean_leve': np.float64(5.188734717905094),\n",
       "  'mean_grave': np.float64(5.250601477208334)},\n",
       " {'gene': 'BNC1',\n",
       "  'pval': np.float64(0.9280758322672268),\n",
       "  'mean_leve': np.float64(1.4190469941909722),\n",
       "  'mean_grave': np.float64(1.4156243855625)},\n",
       " {'gene': 'MPP4',\n",
       "  'pval': np.float64(0.41064019419029607),\n",
       "  'mean_leve': np.float64(2.045786517826389),\n",
       "  'mean_grave': np.float64(2.07775638859375)},\n",
       " {'gene': 'LACE1',\n",
       "  'pval': np.float64(0.67539005099709),\n",
       "  'mean_leve': np.float64(1.989389947295139),\n",
       "  'mean_grave': np.float64(2.011638705359375)},\n",
       " {'gene': 'IDI2',\n",
       "  'pval': np.float64(0.3889120164689085),\n",
       "  'mean_leve': np.float64(5.606372586302083),\n",
       "  'mean_grave': np.float64(5.5426359168124995)},\n",
       " {'gene': 'CYP11B1',\n",
       "  'pval': np.float64(0.851410136311382),\n",
       "  'mean_leve': np.float64(3.111657595364583),\n",
       "  'mean_grave': np.float64(3.0821205043593753)},\n",
       " {'gene': 'TAF8',\n",
       "  'pval': np.float64(0.7227739496996628),\n",
       "  'mean_leve': np.float64(5.023208674080555),\n",
       "  'mean_grave': np.float64(4.982341173575)},\n",
       " {'gene': 'COBL',\n",
       "  'pval': np.float64(0.6289451302995865),\n",
       "  'mean_leve': np.float64(2.6063511338506946),\n",
       "  'mean_grave': np.float64(2.62531507865625)},\n",
       " {'gene': 'SLAMF6',\n",
       "  'pval': np.float64(0.4172229415116322),\n",
       "  'mean_leve': np.float64(7.286619329305555),\n",
       "  'mean_grave': np.float64(7.38142137096875)},\n",
       " {'gene': 'ZSCAN20',\n",
       "  'pval': np.float64(0.5164638931388066),\n",
       "  'mean_leve': np.float64(2.7198089137199073),\n",
       "  'mean_grave': np.float64(2.68183895196875)},\n",
       " {'gene': 'GPBAR1',\n",
       "  'pval': np.float64(0.9972806576311214),\n",
       "  'mean_leve': np.float64(6.5548562301805555),\n",
       "  'mean_grave': np.float64(6.55444196365625)},\n",
       " {'gene': 'RHBDL2',\n",
       "  'pval': np.float64(0.9318299682409064),\n",
       "  'mean_leve': np.float64(4.2224871712314815),\n",
       "  'mean_grave': np.float64(4.237150294145834)},\n",
       " {'gene': 'FRAS1',\n",
       "  'pval': np.float64(0.6156195041958483),\n",
       "  'mean_leve': np.float64(2.580603878342593),\n",
       "  'mean_grave': np.float64(2.606769868609375)},\n",
       " {'gene': 'BRSK1',\n",
       "  'pval': np.float64(0.21319172678400958),\n",
       "  'mean_leve': np.float64(4.3163102465902785),\n",
       "  'mean_grave': np.float64(4.22816874815625)},\n",
       " {'gene': 'CRB2',\n",
       "  'pval': np.float64(0.9822609794563353),\n",
       "  'mean_leve': np.float64(3.9520058395069446),\n",
       "  'mean_grave': np.float64(3.9508596031875003)},\n",
       " {'gene': 'KCNE4',\n",
       "  'pval': np.float64(0.3932209802232859),\n",
       "  'mean_leve': np.float64(1.9677926599953703),\n",
       "  'mean_grave': np.float64(1.9346882305833333)},\n",
       " {'gene': 'CD300LG',\n",
       "  'pval': np.float64(0.9926665463445417),\n",
       "  'mean_leve': np.float64(2.693114242151042),\n",
       "  'mean_grave': np.float64(2.694182473585937)},\n",
       " {'gene': 'SLC34A3',\n",
       "  'pval': np.float64(0.9353557115466755),\n",
       "  'mean_leve': np.float64(3.6896834635833335),\n",
       "  'mean_grave': np.float64(3.6974115341406253)},\n",
       " {'gene': 'CPA6',\n",
       "  'pval': np.float64(0.5159065505098619),\n",
       "  'mean_leve': np.float64(2.0914794848333336),\n",
       "  'mean_grave': np.float64(2.0669050398906252)},\n",
       " {'gene': 'WBP2NL',\n",
       "  'pval': np.float64(0.13420024853234347),\n",
       "  'mean_leve': np.float64(2.3179160341736114),\n",
       "  'mean_grave': np.float64(2.43812228309375)},\n",
       " {'gene': 'HIPK1',\n",
       "  'pval': np.float64(0.9427701924092062),\n",
       "  'mean_leve': np.float64(5.276541958673611),\n",
       "  'mean_grave': np.float64(5.296666405804688)},\n",
       " {'gene': 'MTBP',\n",
       "  'pval': np.float64(0.859152198696393),\n",
       "  'mean_leve': np.float64(2.21081870246875),\n",
       "  'mean_grave': np.float64(2.1999048115859376)},\n",
       " {'gene': 'ACVR1C',\n",
       "  'pval': np.float64(0.3672334870886458),\n",
       "  'mean_leve': np.float64(2.770795872795139),\n",
       "  'mean_grave': np.float64(2.84481715603125)},\n",
       " {'gene': 'TMEM74',\n",
       "  'pval': np.float64(0.9499492395044307),\n",
       "  'mean_leve': np.float64(2.529139828453704),\n",
       "  'mean_grave': np.float64(2.5226157046354167)},\n",
       " {'gene': 'TIGD4',\n",
       "  'pval': np.float64(0.7734591786598297),\n",
       "  'mean_leve': np.float64(2.2931308837500004),\n",
       "  'mean_grave': np.float64(2.279787927359375)}]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8da33210",
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "\n",
    "# Adiciona coluna para -log10(pval)\n",
    "res_df['-log10_pval'] = -np.log10(res_df['pval'])\n",
    "\n",
    "# Define cores: destaque genes significativos\n",
    "colors = np.where((res_df['pval'] < 0.05) & (abs(res_df['log2fc']) > 1), '#A259C6', '#4CB944')\n",
    "\n",
    "fig = go.Figure()\n",
    "\n",
    "fig.add_trace(go.Scatter(\n",
    "    x=res_df['log2fc'],\n",
    "    y=res_df['-log10_pval'],\n",
    "    mode='markers',\n",
    "    marker=dict(\n",
    "        color=colors,\n",
    "        size=8,\n",
    "        line=dict(width=0.5, color='DarkSlateGrey')\n",
    "    ),\n",
    "    text=res_df['gene'],  # hover com nome do gene\n",
    "    hovertemplate='Gene: %{text}<br>log2FC: %{x:.2f}<br>-log10(pval): %{y:.2f}<extra></extra>'\n",
    "))\n",
    "\n",
    "fig.update_layout(\n",
    "    title='Volcano Plot - Expressão Diferencial',\n",
    "    xaxis_title='log2 Fold Change',\n",
    "    yaxis_title='-log10(p-valor)',\n",
    "    template='plotly_white',\n",
    "    width=700,\n",
    "    height=500\n",
    ")\n",
    "\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1ca6a68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'title': ['PBMC_SLE2132_v4'],\n",
       " 'geo_accession': ['GSM3428621'],\n",
       " 'status': ['Public on Oct 15 2018'],\n",
       " 'submission_date': ['Oct 15 2018'],\n",
       " 'last_update_date': ['Oct 15 2018'],\n",
       " 'type': ['RNA'],\n",
       " 'channel_count': ['1'],\n",
       " 'source_name_ch1': ['PBMC_SLE2132_v4'],\n",
       " 'organism_ch1': ['Homo sapiens'],\n",
       " 'taxid_ch1': ['9606'],\n",
       " 'characteristics_ch1': ['disease state: Systemic Lupus Erythematosus',\n",
       "  'patient id: 2132',\n",
       "  'sledai: 2',\n",
       "  'visit date: 2010-12-15',\n",
       "  'tissue: PBMC',\n",
       "  'imputed neutrophil percentage: 56.2'],\n",
       " 'treatment_protocol_ch1': ['Standard.'],\n",
       " 'growth_protocol_ch1': ['Whole Blood samples in collected fom PAXgene tubes.'],\n",
       " 'molecule_ch1': ['total RNA'],\n",
       " 'extract_protocol_ch1': ['The RNA was extracted with Trizol reagent (Life Technologies) and cleaned up with RNA easy (Qiagen) according to the manufacturer’s protocols.'],\n",
       " 'label_ch1': ['Biotin'],\n",
       " 'label_protocol_ch1': ['NuGEN Ovation.'],\n",
       " 'hyb_protocol': ['Standard Affymetrix protcol.'],\n",
       " 'scan_protocol': ['HTAPS scan was used.'],\n",
       " 'description': ['Gene expression microarray from a patient in a visit'],\n",
       " 'data_processing': ['Data was normalized using rma R package.'],\n",
       " 'platform_id': ['GPL13158'],\n",
       " 'contact_name': ['Daniel,,Toro'],\n",
       " 'contact_email': ['danieltorodominguez@gmail.com'],\n",
       " 'contact_department': ['Bioinformatics'],\n",
       " 'contact_institute': ['GENYO'],\n",
       " 'contact_address': ['Avenida de la Ilustracion'],\n",
       " 'contact_city': ['GRANADA'],\n",
       " 'contact_state': ['Spain'],\n",
       " 'contact_zip/postal_code': ['18016'],\n",
       " 'contact_country': ['Spain'],\n",
       " 'supplementary_file': ['ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3428nnn/GSM3428621/suppl/GSM3428621_101613_119686173_169.CEL.gz'],\n",
       " 'series_id': ['GSE121239'],\n",
       " 'data_row_count': ['54715']}"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Visualizacoes uteis\n",
    "\n",
    "# gse = GEOparse.get_GEO(geo=\"GSE121239\", destdir=\"../../data/raw/\", silent=True)\n",
    "# gse.gsms['GSM3428621'].metadata # Metadados pacientes\n",
    "# gse.gsms['GSM3428621'].table # Tabela de genes\n",
    "# gpl.head() # De para de Genes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
